core/src/avm2/vtable.rs

```diff
@@ -64,14 +69,33 @@ impl<'gc> VTable<'gc> {
         VTable(GcCell::allocate(mc, self.0.read().clone()))
     }
 
-    pub fn get_trait(self, name: &Multiname<'gc>) -> Option<Property<'gc>> {
+    pub fn get_trait(self, name: &Multiname<'gc>) -> Option<Property> {
         self.0
             .read()
             .resolved_traits
             .get_for_multiname(name)
             .cloned()
     }
 
+    /// Coerces `value` to the type of the slot with id `slot_id`
+    pub fn coerce_trait_value(
+        &self,
+        slot_id: u32,
+        value: Value<'gc>,
+        activation: &mut Activation<'_, 'gc, '_>,
+    ) -> Result<Value<'gc>, Error> {
+        // Drop the `write()` guard, as 'slot_class.coerce' may need to access this vtable.
+        let mut slot_class =
+            { self.0.write(activation.context.gc_context).slot_classes[slot_id as usize].clone() };
+        let value = slot_class.coerce(activation, value);
+        // Calling coerce modifies `PropertyClass` to cache the class lookup,
+        // so store the new value back in the vtable.
+        {
+            self.0.write(activation.context.gc_context).slot_classes[slot_id as usize] = slot_class;
```

Comment:
will read it more tomorrow, but: we probably only want to do a write if we actually update the entry?